ftdc/custom_format.go

```diff
@@ -0,0 +1,549 @@
+//
+// Because a metric reading is not meaningful without a schema, a file will always start with a
+// schema document. The first byte of a schema document is 0x01 followed by a JSON list of strings
+// and a UNIX newline (0x0a). The JSON strings are "flattened" using a dot to concatenate the map
+// key with the metric name. E.g:
+//
+// ["motor.powerPct", "motor.pos", "gps.lat", "gps.long"]\n
+//
+// Following a schema document will be 0 or more metric documents. A metric reading has one diff bit
+// per reading (i.e: the "size" of the schema). In our example, that's four bits.  A diff bit is set
+// to `0` if the new reading for a given metric is same as the immediately prior reading. A diff bit
+// is set to `1` if the readings differ. Each reading that differs will have one 32-bit float value
+// written as part of this metric reading document.
+//
+// The diff bits immediately follow the metric bit value of 0. In other words, the first byte
+// containing the metric bit is packed/merged with the first (up to seven) diff bits.
+//
+// The initial metric reading immediately following a schema document does not have a diff to
+// compare against. In this case the format assumes a prior value of `0` for each metric. To
+// continue our example, let's consider the first metric reading document.
+//
+// For clarity, the diff bits are described in a binary representation detailing the exact
+// bits. Where bit-0 is the metric bit (defined as 0) and the diff bits 1 through 4 (inclusive) are
+// set to 1. And the numbers written for time/readings are annotated. All numbers are big-endian
+// encoded for no good/benchmarked reason.
+//
+// 0001 1110 <64bit time> <32bit "motor.powerPct"> <32bit "motor.pos"> <32bit "gps.lat"> <32bit "gps.long">
+// 7       0
+//
+// Now let's see what the second datum will look like. First we calculate which metric readings have
+// changed:
+// - motor.powerPct: 0.2 -> 0.2 (no diff)
+// - motor.pos: 5000 -> 5001 (diff)
+// - gps.lat: 40.7128 -> 40.7128 (no diff)
+// - gps.long: -74.0060 -> -74.0061 (diff)
+//
+// Giving us the following encoding:
+//
+// 0001 0100 <64bit time> <32bit "motor.pos"> <32bit "gps.long">
+// 7       0
+//
+// If we now get a new datum that changes the schema (e.g: remove the motor), we can write out a new
+// schema document:
+//
+// 0000 0001 ["gps.lat", "gps.long"]\n
+// 7       0
+//
+// To maybe illuminate the necessity of the schema/metric identifier, when a parser is about to read
+// a new document, it needs to know whether:
+// - to interpret the bytes as json for a new schema, or
+// - as values for a new metric reading
+//
+// A parser can read a single byte and look at the least significant bit to determine which path to
+// take.
+func writeSchema(schema *Schema, output io.Writer) {
+	// New schema byte
+	output.Write([]byte{0x1})
+	encoder := json.NewEncoder(output)
+	// `json.Encoder.Encode` assumes it convenient to append a newline character at the very
```

Comment:
I was being cheeky in criticizing the json library. The library believes it's doing a service for the caller without offering a way to configure that newline away.

Reply:
```
	// Terminate each value with a newline.
	// This makes the output look a little nicer
	// when debugging, and some kind of space
	// is required if the encoded value was a number,
	// so that the reader knows there aren't more
	// digits coming.
	e.WriteByte('\n')
```

Reply:
(I disagree with the assertion that it's legal to call `json.Encode(5)`. "5" is not in javascript object notation).